### PR TITLE
crimson/os/seastore/transaction_manager: move intermediate_key by "remap_offset" when remapping the "back" half of the original pin

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/cached_extent.h [hunk 1]
```diff
@@ -1198,7 +1198,7 @@ class LogicalCachedExtent : public ChildableCachedExtent {
 
   void maybe_set_intermediate_laddr(LBAMapping &mapping) {
     laddr = mapping.is_indirect()
```
Comment (cyx1231st) [inline src/crimson/os/seastore/cached_extent.h:1200]: Let's add comments to the member variable `laddr`:
```// the logical address of the extent, and if shared, it is the intermediate_base, see BtreeLBAMapping comments.```

File: src/crimson/os/seastore/transaction_manager.h [hunk 2]
```diff
@@ -385,17 +385,17 @@ class TransactionManager : public ExtentCallbackInterface {
...
 		   &remaps, intermediate_base, intermediate_key,
-                   original_laddr, original_paddr, original_len](auto) {
+                   original_laddr, original_paddr, original_len](auto) mutable {
           return trans_intr::do_for_each(
             remaps.begin(),
             remaps.end(),
             [this, &t, &original_bptr, &ret,
 	    &count, intermediate_base, intermediate_key,
-	    original_laddr, original_paddr, original_len](auto &remap) {
+	    original_laddr, original_paddr, original_len](auto &remap) mutable {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/transaction_manager.h:398]: How about reverting the above `mutable` keywords and introduce a `remapped_intermediate_key` below ?

**Review by cyx1231st — CHANGES_REQUESTED** [review]
The fix LGTM, except for the comments.

File: src/crimson/os/seastore/transaction_manager.h (hunk 2 above)
Comment (liu-chunmei) [reply src/crimson/os/seastore/transaction_manager.h:398]: I didn't see the captures difference, so which variable is not read only and cause mutable needed?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/transaction_manager.h]: Done.

**Review by cyx1231st — APPROVED** [review]
LGTM with minor nits.

File: src/crimson/os/seastore/cached_extent.h [hunk 3]
```diff
@@ -1197,8 +1197,10 @@ class LogicalCachedExtent : public ChildableCachedExtent {
   }
 
   void maybe_set_intermediate_laddr(LBAMapping &mapping) {
+    // the logical address of the extent, and if shared,
+    // it is the intermediate_base, see BtreeLBAMapping comments.
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/cached_extent.h]: Done.

**Pull request merged**